packages/pkh-webauthn/src/index.ts

```diff
@@ -0,0 +1,247 @@
+  const credential = await credentials.create(opts) as any
+  if (!credential) throw new Error('Empty Credential Response')
+
+  const authenticatorData = getAuthenticatorData(credential.response)
+  const { publicKey } = decodeAuthenticatorData(authenticatorData)
+  for (const selector of session.selectors) {
+    if (typeof selector.seen === 'function') selector.seen(credential.id, publicKey)
+  }
+
+  return {
+    publicKey,
+    did: encodeDIDFromPub(publicKey),
+    credential
+  }
+}
+
+/**
+ * Asks user to sign a random challenge
+ * returns either pk0 or pk1 given correct credential is presented
+ */
+export async function probeAuthenticator(pk0: Uint8Array, pk1: Uint8Array) {
+  const res = await authenticatorSign(randomBytes(32))
+  for (const candidate of res.recovered) {
+    if (!u8a.compare(candidate, pk0)) return pk0
+    if (!u8a.compare(candidate, pk1)) return pk1
+  }
+  throw new Error('DifferentCredentialSelected')
+}
+
+
+export namespace WebauthnAuth {
+  export interface CreateCredentialResult {
+    publicKey: Uint8Array,
+    credential: PublicKeyCredential,
+    did: string
+  }
+  export interface KeySelector {
+    seen?: (credentialId: string, pk: Uint8Array) => void
+    select: (credentialId: string, pk0: Uint8Array, pk1: Uint8Array) => Uint8Array|null
+  }
+  export interface AuthenticatorSession {
+    credentialId?: string // Set if PasskeyProvider was initialized in non-discoverable mode
+    publicKey?: Uint8Array // source for did
+    selectors: KeySelector[]
+  }
+
+  /**
+   * Creates a new session that remembers seen credentials
+   * @param {KeySelector?} selector An object implemeting the KeySelector interface.
+   * @param {string?} credentialId Optionally specify which credential to use for all operations if known ahead of time.
+   */
+  export function createSession (selector?: KeySelector, credentialId?: string): AuthenticatorSession {
+    const selectors: Array<KeySelector> = [new LocalStorageKeySelector()]
+    if (selector) selectors.push(selector)
+    return { credentialId, selectors }
+  }
+
+  export async function getAuthMethod (session: AuthenticatorSession): Promise<AuthMethod> {
+    return async (opts: AuthMethodOpts): Promise<Cacao> => {
+      return createCacao(opts, session)
+    }
+  }
+
+  async function createCacao (opts: AuthMethodOpts, session: AuthenticatorSession): Promise<Cacao> {
+    const now = Date.now()
+    // The public key is not known at pre-sign time.
+    // so we sign a "challenge"-block without Issuer attribute
+    const challenge: Cacao = Object.freeze({
+      h: {
+        t: 'caip122'
+      },
+      p: {
+        domain: globalThis.location.hostname,
+        aud: '' + globalThis.location,
+        iss: '',
+        version: '1',
+        nonce: globalThis.crypto.randomUUID(),
+        resources: [],
+        exp: new Date(now + 7 * 86400000).toISOString(), // 1 week
+        nbf: new Date(now).toISOString(),
+        iat: new Date(now).toISOString(),
```

Comment:
Please default to opts:
https://github.com/ceramicnetwork/js-did/blob/main/packages/cacao/src/cacao.ts#L96-L102

Reply:
Yes. I zoomed past the entire options object i see. thanks for the link.